pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/ColumnPartitionMetadata.java

```diff
@@ -41,77 +36,120 @@
-      @JsonProperty("numPartitions") int numPartitions,
-      @JsonProperty("partitionRanges") @JsonDeserialize(using = PartitionRangesDeserializer.class) List<IntRange> partitionRanges) {
-    super(functionName, numPartitions);
-    _partitionRanges = partitionRanges;
+  public ColumnPartitionMetadata(String functionName, int numPartitions, List<Integer> partitions) {
+    _functionName = functionName;
+    _numPartitions = numPartitions;
+    _partitions = partitions;
   }
 
-  /**
-   * Returns the list of partition ranges.
-   *
-   * @return List of partition ranges.
-   */
-  @JsonSerialize(using = PartitionRangesSerializer.class)
-  public List<IntRange> getPartitionRanges() {
-    return _partitionRanges;
+  public String getFunctionName() {
+    return _functionName;
+  }
+
+  public int getNumPartitions() {
+    return _numPartitions;
+  }
+
+  public List<Integer> getPartitions() {
+    return _partitions;
   }
 
   @Override
-  public boolean equals(Object o) {
-    if (this == o) {
+  public boolean equals(Object obj) {
+    if (this == obj) {
       return true;
     }
-    if (o == null || getClass() != o.getClass()) {
-      return false;
+    if (obj instanceof ColumnPartitionMetadata) {
+      ColumnPartitionMetadata that = (ColumnPartitionMetadata) obj;
+      return _functionName.equals(that._functionName) && _numPartitions == that._numPartitions && _partitions
+          .equals(that._partitions);
     }
-
-    ColumnPartitionMetadata that = (ColumnPartitionMetadata) o;
-    return super.equals(that) && Objects.equals(_partitionRanges, that._partitionRanges);
+    return false;
   }
 
   @Override
   public int hashCode() {
-    int hashCode = _partitionRanges != null ? _partitionRanges.hashCode() : 0;
+    int hashCode = _partitions != null ? _partitions.hashCode() : 0;
     return EqualityUtils.hashCodeOf(super.hashCode(), hashCode);
   }
 
   /**
-   * Custom Json serializer for list of IntRange's.
+   * Helper method to extract partitions from configuration.
+   * <p>
+   * There are two format of partition strings:
+   * <ul>
+   *   <li>Integer format: e.g. {@code "0"}</li>
+   *   <li>Range format (legacy): e.g. {@code "[0 0]"}</li>
+   * </ul>
    */
-  public static class PartitionRangesSerializer extends JsonSerializer<List<IntRange>> {
-
-    @Override
-    public void serialize(List<IntRange> value, JsonGenerator jsonGenerator, SerializerProvider provider)
-        throws IOException {
-      jsonGenerator.writeString(ColumnPartitionConfig.rangesToString(value));
+  public static List<Integer> extractPartitions(List partitionList) {
+    int numPartitions = partitionList.size();
+    List<Integer> partitions = new ArrayList<>(numPartitions);
+    for (Object o : partitionList) {
+      String partitionString = o.toString();
+      if (partitionString.charAt(0) == '[') {
+        // Range format
+        partitions.add(Integer.parseInt(partitionString.substring(1, partitionString.indexOf(' '))));
```

Comment:
I think that we need to check both numbers in the range. When a segment includes partition 0-9, I have seen something like `partitionRanges: [0 9]`, not `[0 0], [1 1] ... [9 9]`